devito/finite_differences/differentiable.py

```diff
@@ -411,17 +450,46 @@ def _gather_for_diff(self):
 class Pow(DifferentiableOp, sympy.Pow):
     _fd_priority = 0
     __sympy_class__ = sympy.Pow
-    __new__ = DifferentiableOp.__new__
 
 
 class Mod(DifferentiableOp, sympy.Mod):
     __sympy_class__ = sympy.Mod
-    __new__ = DifferentiableOp.__new__
 
 
 class EvalDerivative(DifferentiableOp, sympy.Add):
-    __sympy_class__ = sympy.Add
-    __new__ = DifferentiableOp.__new__
+
+    is_commutative = True
+
+    def __new__(cls, *args, base=None, **kwargs):
+        kwargs['evaluate'] = False
+
+        # a+0 -> a
+        args = [i for i in args if i != 0]
+
+        # Reorder for homogeneity with pure SymPy types
+        _addsort(args)
+
+        obj = super().__new__(cls, *args, **kwargs)
+
+        try:
+            obj.base = base
```

Comment:
CIRE won't optimize EvalDerivatives if the base is a pure Function, such as `u.dx`, simply because there would be nothing to optimize. It needs to be an expression to be worth optimizing, ie `expr.dx` . CTRL+F for `.base` and you'll see where it's used in `aliases.py`